pkg/proxy/iptables/proxier_test.go

```diff
@@ -4957,17 +5075,28 @@ COMMIT
 
 			fp.OnEndpointSliceAdd(testcase.endpointslice)
 			fp.syncProxyRules()
-			t.Log(fp.iptablesData.String())
-			assertIPTablesRulesEqual(t, testcase.expectedIPTables, fp.iptablesData.String())
```

Comment:
I feel we loose coverage here, since the test will pass with the rules that are not exercised by the trace, right?

Reply:
We're not losing signal, we're losing noise. The `assertIPTablesRulesEqual` tests were asserting too much; they made sure that the things we cared about were right, but they also made sure that a lot of things we _didn't_ care about were exactly one particular way, even though we didn't care whether they were that way or not. And this made it really really hard to figure out whether we were actually asserting what we meant to be asserting.

eg, https://github.com/kubernetes/kubernetes/pull/106497/commits/449a8d1f482622c46b87d3f9167afa6a89ee2cef is a major reorganization of the iptables rules, and you can see in the diff to `TestOverallIPTablesRulesWithMultipleServices` that a whole bunch of rules get changed as a result... But what do those changes _mean_? It's really hard to say.

But looking at the rest of the diff, you can see that even though there are a ton of changes to the generated rules, the actual observed behavior only changes in 4 of the trace tests, 2 of which had comments pointing out that they were clearly broken before. (And the other 2 cover the same cases and were also wrong, though they didn't have comments pointing that out.)

And most importantly, the current version of that commit is different from [the original version](https://github.com/kubernetes/kubernetes/commit/34d73d88d2b8797e009b94285bce16ecdcfdeab4), which turned out to be broken. In theory, you should be able to look at the changes to the `assertIPTablesRulesEqual` tests in that commit and tell that the changes are wrong. But neither Tim nor I was able to do that. Whereas once I added the trace tests, it was clear that the new rules were broken:

```
    --- FAIL: TestOverallIPTablesRulesWithMultipleServices/node_to_clusterIP (0.00s)
        rules_test.go:1014: wrong masq: expected true got false
    --- FAIL: TestOverallIPTablesRulesWithMultipleServices/external_to_clusterIP (0.00s)
        rules_test.go:1011: wrong output: expected "10.180.0.2:80" got "DROP"
    --- FAIL: TestOverallIPTablesRulesWithMultipleServices/node_to_nodePort (0.00s)
        rules_test.go:1014: wrong masq: expected true got false
    --- FAIL: TestOverallIPTablesRulesWithMultipleServices/node_to_externalIP (0.00s)
        rules_test.go:1014: wrong masq: expected true got false
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/cluster_IP_to_non-local_endpoint (0.00s)
        rules_test.go:1011: wrong output: expected "10.180.0.1:80" got "10.180.2.1:80"
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/cluster_IP_to_local_endpoint (0.00s)
        rules_test.go:1011: wrong output: expected "10.180.2.1:80" got ""
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/external_to_NodePort (0.00s)
        rules_test.go:1011: wrong output: expected "10.180.2.1:80" got "10.180.0.1:80"
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/external_to_NodePort,_no_second_endpoint (0.00s)
        rules_test.go:1011: wrong output: expected "" got "10.180.2.1:80"
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/pod_to_NodePort (0.00s)
        rules_test.go:1011: wrong output: expected "10.180.2.1:80" got "10.180.0.1:80"
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/pod_to_NodePort,_no_second_endpoint (0.00s)
        rules_test.go:1011: wrong output: expected "" got "10.180.2.1:80"
...
```

Reply:
ic, so you did this based on your experience and I was judging from the top of my head, I retire my comment then since seems to be based on a false assumption

Reply:
To be fair, I didn't TRY to coninvce myself it was right, but your point stands.

I think the right answer is a combination of small tests which target specific functionality and use the trace *and* dump-compare approaches and larger tests which just use the traces.

Like, concretely, the comments ARE part of the tested surface, but the traces throw them out.  I am sure there are other aspects that are similar.